### PR TITLE
Unifying blas interface

### PR DESCRIPTION
**Pull request by mehdi-goli**

### CODE REVIEW

File: include/interface/blas1_interface_sycl.hpp [hunk 1]
```diff
@@ -94,222 +101,250 @@ void _copy(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
...
 #ifdef VERBOSE
-  my_vx.printH("VX");
-  my_vy.printH("VY");
+  my_rs.printH("VR");
 #endif  //  VERBOSE
 }
 
 /**
- * \brief SCAL scales a vector by a constant
+ * \briefCompute the inner product of two vectors with extended
```
Comment (Ruyk) [inline include/interface/blas1_interface_sycl.hpp]: space between brief and compute

File: include/interface/blas1_interface_sycl.hpp [hunk 2]
```diff
@@ -94,222 +101,250 @@ void _copy(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
...
-T _dot(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
-       int _incx, vector_view<T, ContainerT> _vy, int _incy) {
-  auto my_vx = vector_view<T, ContainerT>(_vx, _vx.getDisp(), _incx, _N);
-  auto my_vy = vector_view<T, ContainerT>(_vy, _vy.getDisp(), _incy, _N);
+template <typename ExecutorType, typename T, typename I>
+void _iamax(Executor<ExecutorType> &ex, int _N, T *_vx, int _incx, I *_rs) {
+  using RHS =
+      vector_view<T, typename Executor<ExecutorType>::template ContainerT<T>>;
+  using RHS1 =
+      vector_view<I, typename Executor<ExecutorType>::template ContainerT<I>>;
```
Comment (Ruyk) [inline include/interface/blas1_interface_sycl.hpp]: explain RHS1 vs RHS? can you put more descriptive names?

File: include/interface/blas1_interface_sycl.hpp [hunk 3]
```diff
@@ -94,222 +101,250 @@ void _copy(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
...
+  auto vx_container = ex.get_buffer(_vx);
+  RHS my_vx(vx_container, ex.get_offset(_vx), _incx, _N);
+  auto rs_container = ex.get_buffer(_rs);
+  RHS1 my_rs(rs_container, ex.get_offset(_rs), 1, 1);
 #ifdef VERBOSE
   my_vx.printH("VX");
-  my_vy.printH("VY");
 #endif  //  VERBOSE
-  auto prdOp = make_op<BinaryOp, prdOp2_struct>(my_vx, my_vy);
+  size_t localSize = 256, nWG = 512;
```
Comment (Ruyk) [inline include/interface/blas1_interface_sycl.hpp]: is this set in stone? mark it const and explain how they can be set

Comment (Ruyk) [reply include/interface/blas1_interface_sycl.hpp]: create a github issue to make this device-selectable

File: include/interface/blas1_interface_sycl.hpp [hunk 4]
```diff
@@ -94,222 +101,250 @@ void _copy(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
...
-      make_addAssignReduction(val1, prdOp, localSize, nWG * localSize);
-  ex.execute(assignOp1);
+/**
+ * \brief ICAMAX finds the index of the first element having maximum
+ * @param _vx  VectorView
+ * @param _incx Increment in X axis
+ */
+template <typename ExecutorType, typename T>
+size_t _iamax(Executor<ExecutorType> &ex, int _N, T *_vx, int _incx) {
+  std::vector<IndVal<T>> rsT(1);
```
Comment (Ruyk) [inline include/interface/blas1_interface_sycl.hpp]: IndVal needs a more descriptive name

File: include/interface/blas1_interface_sycl.hpp [hunk 5]
```diff
@@ -94,222 +101,250 @@ void _copy(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
...
-  auto nWG = 512;
-  auto assignOp1 =
-      make_addAssignReduction(my_rs, prdOp, localSize, localSize * nWG);
-  ex.reduce(assignOp1);
-  auto sqrtOp = make_op<UnaryOp, sqtOp1_struct>(my_rs);
-  auto assignOp2 = make_op<Assign>(my_rs, sqrtOp);
-  ex.execute(assignOp2);
 #ifdef VERBOSE
-  my_rs.printH("VR");
+//  std::cout << "val = " << std::sqrt(val2.eval(0)) << std::endl;
```
Comment (Ruyk) [inline include/interface/blas1_interface_sycl.hpp]: commented code

File: include/interface/blas2_interface_sycl.hpp [hunk 6]
```diff
@@ -43,21 +43,29 @@ namespace blas {
...
+
+  size_t M = (_Trans == 'n') ? _M : _N;
+  size_t N = (_Trans == 'n') ? _N : _M;
+  auto _mA_container = ex.get_buffer(_mA);
+  using RHS =
+      matrix_view<T, typename Executor<ExecutorType>::template ContainerT<T> >;
+
+  RHS my_mA(_mA_container, M, N, accessOpr, _lda, ex.get_offset(_mA));
+  using RHS1 =
+      vector_view<T, typename Executor<ExecutorType>::template ContainerT<T> >;
```
Comment (Ruyk) [inline include/interface/blas2_interface_sycl.hpp]: naming

File: include/interface/blas2_interface_sycl.hpp [hunk 7]
```diff
@@ -43,21 +43,29 @@ namespace blas {
...
-  auto my_vy = vector_view<T, ContainerT>(_vy, _vy.getDisp(), _incy, M);
+  int accessOpr = (_Trans == 'n');
+
+  size_t M = (_Trans == 'n') ? _M : _N;
+  size_t N = (_Trans == 'n') ? _N : _M;
+  auto _mA_container = ex.get_buffer(_mA);
+  using RHS =
+      matrix_view<T, typename Executor<ExecutorType>::template ContainerT<T> >;
+
+  RHS my_mA(_mA_container, M, N, accessOpr, _lda, ex.get_offset(_mA));
```
Comment (Ruyk) [inline include/interface/blas2_interface_sycl.hpp]: use brace initializer list

File: include/interface/blas2_interface_sycl.hpp [hunk 8]
```diff
@@ -108,16 +118,20 @@ void _gemv(Executor<ExecutorType> ex, std::string _Trans, size_t _M, size_t _N,
     std::cout << "COLS_2" << std::endl;
 #endif  // VERBOSE
     auto nThr = 2;
-    ContainerT valT1(nThr * M);
-    auto mat1 = matrix_view<T, ContainerT>(valT1, M, nThr);
+    // ContainerT valT1(nThr * M);
```
Comment (Ruyk) [inline include/interface/blas2_interface_sycl.hpp]: commented code

File: include/interface/blas2_interface_sycl.hpp [hunk 9]
```diff
@@ -108,16 +118,20 @@ void _gemv(Executor<ExecutorType> ex, std::string _Trans, size_t _M, size_t _N,
...
     auto scalOp1 = make_op<ScalarOp, prdOp2_struct>(_beta, my_vy);
+    RHS mat1(valT1, M, nThr);
 #ifdef BLAS_EXPERIMENTAL
-    auto val1 = vector_view<T, ContainerT>(valT1, 0, 1, nThr * M);
-    auto mat1 = matrix_view<T, ContainerT>(valT1, M, nThr);
-    auto scalOp1 = make_op<ScalarOp, prdOp2_struct>(_beta, my_vy);
+    auto val1 = RHS1(valT1, 0, 1, nThr * M);
+    //  auto mat1 = RHS(valT1, M, nThr); // has been created outside
+    //    auto scalOp1 = make_op<ScalarOp, prdOp2_struct>(_beta, my_vy); it has
+    //    been created outside
```
Comment (Ruyk) [inline include/interface/blas2_interface_sycl.hpp]: commented code

File: include/operations/blas3_trees.hpp [hunk 10]
```diff
@@ -1,7 +1,6 @@
 /***************************************************************************
- *
  *  @license
- *  Copyright (C) 2016 Codeplay Software Limited
+ *  Copyright (C) 2017 Codeplay Software Limited
```
Comment (Ruyk) [inline include/operations/blas3_trees.hpp]: Remove date in Copyright

File: include/operations/blas3_trees.hpp [hunk 11]
```diff
@@ -19,63 +18,726 @@
...
+          typename T>
+class ReferenceGemmFactory {
+ public:
+  using value_type = T;
+
+  static const int version = 2;
+  static const int wg_size = WgSize;
+  static const bool trans_a = TransA;
+  static const bool trans_b = TransB;
+  static const int scratch_size = 0;
```
Comment (Ruyk) [inline include/operations/blas3_trees.hpp]: static const should be constexpr

File: include/queue/pointer_alias.hpp [hunk 12]
```diff
@@ -17,35 +17,43 @@
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  *
- *  SYCL-BLAS: BLAS implementation using SYCL
+ *  Codeplay's ComputeCpp SDK
```
Comment (Ruyk) [inline include/queue/pointer_alias.hpp]: this is not the SDK

File: include/queue/pointer_mapper.hpp [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  virtual_ptr.hpp
+ *
```
Comment (Ruyk) [inline include/queue/pointer_mapper.hpp]: specify commit id of this file from the original repo

File: include/queue/queue_sycl.hpp [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+ public:
+  enum device_type { UNSUPPORTED_DEVICE, INTELGPU, AMDGPU };
+
+  explicit Queue_Interface(cl::sycl::queue q)
+      : q_(q), pointer_mapper_owner(true) {}
+  // FIXME : The assignment constructor for pointer mapper is deleted
+  /*  Queue_Interface(cl::sycl::queue q,
+                    cl::sycl::codeplay::PointerMapper& pointer_mapper_)
+        : q_(q), pointer_mapper(pointer_mapper_), pointer_mapper_owner(false)
+     {}*/
```
Comment (Ruyk) [inline include/queue/queue_sycl.hpp]: create github issue

File: include/queue/queue_sycl.hpp [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+  device_type get_device_type() {
```
Comment (Ruyk) [inline include/queue/queue_sycl.hpp]: this can be const

File: include/queue/queue_sycl.hpp [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+      @param src is the device pointer we want to copy from.
+      @param dst is the host pointer we want to copy to.
+      @param size is the number of elements to be copied
+  */
+  template <typename T>
+  void copy_to_host(T *src, T *dst, size_t size) {
+    auto buffer = pointer_mapper.get_buffer(static_cast<void *>(src));
+    auto offset = pointer_mapper.get_offset(static_cast<void *>(src));
+    q_.wait();  // FIXME: we should not have that when the size of the buffer is
+                // 1 there is an issue
```
Comment (Ruyk) [inline include/queue/queue_sycl.hpp]: create an issue on computecpp service desk if there is a bug in the package. Check with the latest one

File: include/queue/queue_sycl.hpp [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+                                     cl::sycl::access::target::global_buffer>(
+              cgh, cl::sycl::range<1>(size * sizeof(T)),
+              cl::sycl::id<1>(offset));
+      cgh.copy(read_acc, static_cast<generic_buffer_data_type *>(
+                             static_cast<void *>(dst)));
+    });
+    q_.wait();
+  }
+};  // namespace blas
+}  // namespace blas
```
Comment (Ruyk) [inline include/queue/queue_sycl.hpp]: cannot be 2 namespace blas, something is fishy there

File: include/queue/queue_sycl.hpp [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+#endif  //
```
Comment (Ruyk) [inline include/queue/queue_sycl.hpp]: missing macro comment

File: test/unittest/CMakeLists.txt [hunk 19]
```diff
@@ -19,6 +19,9 @@ file(GLOB SYCL_UNITTEST_SRCS
   ${SYCLBLAS_UNITTEST}/blas1_rotg_test.cpp
   ${SYCLBLAS_UNITTEST}/blas1_iamax_test.cpp
   ${SYCLBLAS_UNITTEST}/blas1_iamin_test.cpp
+  #${SYCLBLAS_UNITTEST}/blas2_gemv_test.cpp
+  #${SYCLBLAS_UNITTEST}/blas2_ger_test.cpp
+  #${SYCLBLAS_UNITTEST}/blas3_gemm_test.cpp
```
Comment (Ruyk) [inline test/unittest/CMakeLists.txt]: commented on purpose?

File: include/interface/blas1_interface_sycl.hpp (hunk 1 above)
Comment (mehdi-goli) [reply include/interface/blas1_interface_sycl.hpp]: removed

File: include/interface/blas1_interface_sycl.hpp (hunk 2 above)
Comment (mehdi-goli) [reply include/interface/blas1_interface_sycl.hpp]: I have renamed them to the reasonable name

File: include/interface/blas1_interface_sycl.hpp (hunk 3 above)
Comment (mehdi-goli) [reply include/interface/blas1_interface_sycl.hpp]: I have put  TODO to fix it on the code. I have also created the issue on the github as well.

File: include/interface/blas1_interface_sycl.hpp (hunk 4 above)
Comment (mehdi-goli) [reply include/interface/blas1_interface_sycl.hpp]: Renamed to IndexValueTuple

File: include/interface/blas1_interface_sycl.hpp (hunk 5 above)
Comment (mehdi-goli) [reply include/interface/blas1_interface_sycl.hpp]: removed

File: include/interface/blas2_interface_sycl.hpp (hunk 6 above)
Comment (mehdi-goli) [reply include/interface/blas2_interface_sycl.hpp]: renamed to VectorView

File: include/interface/blas2_interface_sycl.hpp (hunk 7 above)
Comment (mehdi-goli) [reply include/interface/blas2_interface_sycl.hpp]: changed to {} instead of ()

File: include/interface/blas2_interface_sycl.hpp (hunk 8 above)
Comment (mehdi-goli) [reply include/interface/blas2_interface_sycl.hpp]: Removed

File: include/interface/blas2_interface_sycl.hpp (hunk 9 above)
Comment (mehdi-goli) [reply include/interface/blas2_interface_sycl.hpp]: removed

File: include/operations/blas3_trees.hpp (hunk 10 above)
Comment (mehdi-goli) [reply include/operations/blas3_trees.hpp]: removed

File: include/queue/queue_sycl.hpp (hunk 14 above)
Comment (mehdi-goli) [reply include/queue/queue_sycl.hpp]: created and the commented code has been removed

File: include/queue/queue_sycl.hpp (hunk 15 above)
Comment (mehdi-goli) [reply include/queue/queue_sycl.hpp]: const has been added

File: include/queue/queue_sycl.hpp (hunk 16 above)
Comment (mehdi-goli) [reply include/queue/queue_sycl.hpp]: created.

File: include/queue/queue_sycl.hpp (hunk 17 above)
Comment (mehdi-goli) [reply include/queue/queue_sycl.hpp]: Fixed

File: include/queue/queue_sycl.hpp (hunk 18 above)
Comment (mehdi-goli) [reply include/queue/queue_sycl.hpp]: Fixed

File: include/views/view_sycl.hpp [hunk 20]
```diff
@@ -510,10 +512,11 @@ struct matrix_view<ScalarT, bufferT<ScalarT>> {
 #endif  // __SYCL_DEVICE_ONLY__
     ScalarT retVal;
     {
+      // however for the host accessor it can be used as we did not use range
+      // accessor here
       auto hostPtr =
-          data_.template get_access<cl::sycl::access::mode::read_write,
-                                    cl::sycl::access::target::host_buffer>();
-      retVal = hostPtr[ind];
+          data_.template get_access<cl::sycl::access::mode::read_write>();
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply include/views/view_sycl.hpp]: Fixed

File: test/unittest/CMakeLists.txt (hunk 19 above)
Comment (mehdi-goli) [reply test/unittest/CMakeLists.txt]: Removed

File: test/unittest/blas1_copy_test.cpp [hunk 21]
```diff
@@ -69,4 +68,7 @@ TYPED_TEST(BLAS1_Test, copy_test) {
       ASSERT_EQ(0, vY[i]);
     }
   }
+
+  ex.template deallocate<ScalarT>(gpu_vX);
+  ex.template deallocate<ScalarT>(gpu_vY);
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply test/unittest/blas1_copy_test.cpp]: I will apply it as a new pull request with new c++ interface in a separate pull request. I have created an issue for it .

File: include/queue/pointer_alias.hpp (hunk 12 above)
Comment (mehdi-goli) [reply include/queue/pointer_alias.hpp]: Removed

File: include/queue/pointer_mapper.hpp (hunk 13 above)
Comment (mehdi-goli) [reply include/queue/pointer_mapper.hpp]: Added

File: test/exprtest/blas1_iface_test.cpp [hunk 22]
```diff
@@ -23,25 +23,25 @@
...
   using ScalarT = typename TypeParam::scalar_t;
   using ExecutorType = typename TypeParam::executor_t;
-  using TestClass = BLAS1_Test<TypeParam>;
+  using TestClass = BLAS_Test<TypeParam>;
   using test = class interface1_test;
 
-  size_t size = TestClass::template test_size<test>();
-  size_t strd = TestClass::template test_strd<test>();
+  size_t size = 25;  // TestClass::template test_size<test>();
+  size_t strd = 1;   // TestClass::template test_strd<test>();
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply test/exprtest/blas1_iface_test.cpp]: Fixed

**Pull request merged**